File: Backend/routes/stats.py
from flask import Blueprint, jsonify
from services.db import fetch_all

bp = Blueprint('stats', __name__)

from services.db import fetch_one, execute_query
import logging

def registrar_evento(tipo, nombre_host):
    try:
        host = fetch_one("SELECT id_host FROM hosts WHERE nombre = %s", (nombre_host,))
        if host:
            execute_query(
                "INSERT INTO estadisticas (id_host, tipo, timestamp) VALUES (%s, %s, NOW())",
                (host['id_host'], tipo)
            )
        else:
            logging.warning(f"No se encontró el host '{nombre_host}' para registrar evento '{tipo}'")
    except Exception as e:
        logging.error(f"Error al registrar evento '{tipo}' para host '{nombre_host}': {str(e)}")


@bp.route('/dashboard', methods=['GET'])
def get_dashboard_stats():
    try:
        # Clientes por servidor
        clientes_por_servidor = fetch_all("""
            SELECT servidor_asignado AS servidor, COUNT(*) AS total_clientes
            FROM clientes_activos
            GROUP BY servidor_asignado;
        """)

        # Total de transmisiones 
        transmisiones_activas = fetch_all("""
            SELECT COUNT(DISTINCT ip_destino) AS total_transmisiones
            FROM servidores_vlc_activos
            WHERE status = 'activo';
        """)[0]['total_transmisiones']

        # Clientes activos totales
        total_clientes = fetch_all("""
            SELECT COUNT(*) AS total_clientes
            FROM clientes_activos;
        """)[0]['total_clientes']

        # Carga vs peso
        carga_vs_peso = fetch_all("""
            SELECT s.host_name AS servidor, s.server_weight AS peso_configurado,
                   COUNT(c.host_cliente) AS clientes_asignados
            FROM servidores_vlc_activos s
            LEFT JOIN clientes_activos c ON c.servidor_asignado = s.host_name
            WHERE s.status = 'activo'
            GROUP BY s.host_name, s.server_weight;
        """)

        # Últimos eventos 
        ultimos_eventos = fetch_all("""
            SELECT h.nombre AS host, e.tipo AS tipo_evento, e.timestamp
            FROM estadisticas e
            JOIN hosts h ON e.id_host = h.id_host
            ORDER BY e.timestamp DESC
            LIMIT 10;
        """)

        # Total de flujos multicast activos 
        flujos_multicast = fetch_all("""
            SELECT ip_destino AS grupo, COUNT(*) AS total_puertos
            FROM clientes_activos
            GROUP BY ip_destino;
        """)

        return jsonify({
            "clientes_por_servidor": clientes_por_servidor,
            "transmisiones_activas": transmisiones_activas,
            "total_clientes": total_clientes,
            "carga_vs_peso": carga_vs_peso,
            "ultimos_eventos": ultimos_eventos,
            "flujos_multicast": flujos_multicast
        }), 200

    except Exception as e:
        return jsonify({"error": f"Error al generar estadísticas: {str(e)}"}), 500

@bp.route('/combined_stats', methods=['GET'])
def get_combined_stats():
    try:
        # Fetch combined data from both rutas_ping and latencias
        combined_stats = fetch_all("""
            SELECT
                l.host_origen,
                l.host_destino,
                r.descripcion AS ruta,
                r.algoritmo_enrutamiento,
                l.rtt_ms AS rtt,
                l.jitter_ms AS jitter
            FROM rutas_ping r
            JOIN latencias l ON r.id_ruta = l.id_ruta;
        """)
        
        return jsonify(combined_stats), 200
    
    except Exception as e:
        return jsonify({"error": f"Error al generar estadísticas combinadas: {str(e)}"}), 500

@bp.route('/comparar_algoritmos', methods=['GET'])
def comparar_algoritmos():
    try:
        # Obtener las métricas de RTT y Jitter para cada algoritmo
        metrics = fetch_all("""
            SELECT r.algoritmo_enrutamiento, 
                   AVG(l.rtt_ms) AS avg_rtt, 
                   AVG(l.jitter_ms) AS avg_jitter
            FROM rutas_ping r
            JOIN latencias l ON r.id_ruta = l.id_ruta
            WHERE r.algoritmo_enrutamiento IN ('dijkstra', 'shortest_path')
            GROUP BY r.algoritmo_enrutamiento;
        """)

        # Preparar los datos para el gráfico
        data = {
            "dijkstra": {
                "avg_rtt": None,
                "avg_jitter": None
            },
            "shortest_path": {
                "avg_rtt": None,
                "avg_jitter": None
            }
        }

        for metric in metrics:
            if metric['algoritmo_enrutamiento'] == 'dijkstra':
                data['dijkstra']['avg_rtt'] = metric['avg_rtt']
                data['dijkstra']['avg_jitter'] = metric['avg_jitter']
            elif metric['algoritmo_enrutamiento'] == 'shortest_path':
                data['shortest_path']['avg_rtt'] = metric['avg_rtt']
                data['shortest_path']['avg_jitter'] = metric['avg_jitter']

        return jsonify(data), 200

    except Exception as e:
        return jsonify({"error": f"Error al generar la comparación: {str(e)}"}), 500

File: Backend/static/js/stats.js
async function cargarEstadisticas() {
  try {
    const res = await fetch("/stats/dashboard");
    const stats = await res.json();

    // Tarjetas resumen
    document.getElementById("total-clientes").textContent = stats.total_clientes;
    document.getElementById("total-transmisiones").textContent = stats.transmisiones_activas;
    document.getElementById("total-grupos").textContent = stats.flujos_multicast.length;

    // Gráfico de barras - Clientes por servidor
    const grafico = document.getElementById("grafico-clientes");
    grafico.innerHTML = ""; 

    // Preparar los datos para el gráfico
    const labels = stats.clientes_por_servidor.map(item => item.servidor);
    const clientesData = stats.clientes_por_servidor.map(item => item.total_clientes);

    // Verificar si ya existe un gráfico y destruirlo antes de crear uno nuevo
    if (window.myChart) {
      window.myChart.destroy(); 
    }

    // Crear el gráfico
    const ctx = grafico.getContext("2d");
    window.myChart = new Chart(ctx, {
      type: 'bar',
      data: {
        labels: labels,
        datasets: [
          {
            label: 'Clientes Asociados',
            data: clientesData,
            backgroundColor: 'rgba(54, 162, 235, 0.5)',
            borderColor: 'rgba(54, 162, 235, 1)',
            borderWidth: 1
          }
        ]
      },
      options: {
        scales: {
          y: {
            beginAtZero: true,
            ticks: {
              stepSize: 1
            }
          }
        },
        responsive: true,
        plugins: {
          title: {
            display: true,
            text: 'Clientes Asociados a Servidores Activos'
          },
          legend: {
            display: false
          }
        }
      }
    });

    // Tabla carga vs peso 
    const tablaCarga = document.getElementById("tabla-carga");
    tablaCarga.innerHTML = "";
    stats.carga_vs_peso.forEach(item => {
      const row = `<tr>
        <td>${item.servidor}</td>
        <td>${item.clientes_asignados}</td>
        <td>${item.peso_configurado}</td>
      </tr>`;
      tablaCarga.innerHTML += row;
    });

    // Lista de eventos
    const listaEventos = document.getElementById("lista-eventos");
    listaEventos.innerHTML = "";
    stats.ultimos_eventos.forEach(ev => {
      const item = document.createElement("li");
      const fecha = new Date(ev.timestamp).toLocaleString();
      item.textContent = `${ev.host}: ${ev.tipo_evento} (${fecha})`;
      listaEventos.appendChild(item);
    });
  } catch (err) {
    console.error("Error cargando estadísticas:", err);
  }
}


async function cargarEstadisticasConRutas() {
  try {
    const res = await fetch("/stats/combined_stats");
    const stats = await res.json();

    // Tabla para mostrar las estadísticas combinadas
    const tablaEstadisticas = document.getElementById("tabla-estadisticas");
    tablaEstadisticas.innerHTML = "";  

    // Encabezados de la tabla
    tablaEstadisticas.innerHTML = `
      <thead>
        <tr>
          <th class="py-2 px-4 border-b bg-gray-50 text-left text-xs font-semibold text-gray-600 uppercase">Host Origen</th>
          <th class="py-2 px-4 border-b bg-gray-50 text-left text-xs font-semibold text-gray-600 uppercase">Host Destino</th>
          <th class="py-2 px-4 border-b bg-gray-50 text-left text-xs font-semibold text-gray-600 uppercase">Ruta</th>
          <th class="py-2 px-4 border-b bg-gray-50 text-left text-xs font-semibold text-gray-600 uppercase">Algoritmo</th>
          <th class="py-2 px-4 border-b bg-gray-50 text-left text-xs font-semibold text-gray-600 uppercase">RTT (ms)</th>
          <th class="py-2 px-4 border-b bg-gray-50 text-left text-xs font-semibold text-gray-600 uppercase">Jitter (ms)</th>
        </tr>
      </thead>
      <tbody>
    `;

    // Filas de la tabla
    stats.forEach(item => {
      const row = `
        <tr>
          <td class="py-2 px-4">${item.host_origen}</td>
          <td class="py-2 px-4">${item.host_destino}</td>
          <td class="py-2 px-4">${item.ruta}</td>
          <td class="py-2 px-4">${item.algoritmo_enrutamiento}</td>
          <td class="py-2 px-4">${item.rtt}</td>
          <td class="py-2 px-4">${item.jitter || "N/A"}</td>
        </tr>
      `;
      tablaEstadisticas.innerHTML += row;
    });

    tablaEstadisticas.innerHTML += "</tbody>";

  } catch (err) {
    console.error("Error cargando estadísticas combinadas:", err);
  }
}

async function cargarComparacionAlgoritmos() {
  try {
    const res = await fetch("/stats/comparar_algoritmos");
    const data = await res.json();

    // Obtener los datos para el gráfico
    const labels = ['RTT (ms)', 'Jitter (ms)'];
    const dijkstraData = [data.dijkstra.avg_rtt, data.dijkstra.avg_jitter];
    const shortestPathData = [data.shortest_path.avg_rtt, data.shortest_path.avg_jitter];

    // Crear el gráfico
    const ctx = document.getElementById("comparacionAlgoritmosChart").getContext("2d");
    new Chart(ctx, {
      type: 'bar',
      data: {
        labels: labels,
        datasets: [
          {
            label: 'Dijkstra',
            data: dijkstraData,
            backgroundColor: 'rgba(54, 162, 235, 0.5)',
            borderColor: 'rgba(54, 162, 235, 1)',
            borderWidth: 1
          },
          {
            label: 'Shortest Path',
            data: shortestPathData,
            backgroundColor: 'rgba(255, 99, 132, 0.5)',
            borderColor: 'rgba(255, 99, 132, 1)',
            borderWidth: 1
          }
        ]
      },
      options: {
        scales: {
          y: {
            beginAtZero: true,
            ticks: {
              stepSize: 1
            }
          }
        },
        responsive: true,
        plugins: {
          title: {
            display: true,
            text: 'Comparación de Algoritmos (RTT y Jitter)'
          },
          legend: {
            position: 'top'
          }
        }
      }
    });
  } catch (err) {
    console.error("Error al cargar la comparación de algoritmos:", err);
  }
}

document.addEventListener("DOMContentLoaded", cargarComparacionAlgoritmos);
document.getElementById("tabla-estadisticas").innerHTML = "<p>Cargando estadísticas...</p>";
cargarEstadisticasConRutas();


document.addEventListener("DOMContentLoaded", cargarEstadisticas);
